### PR TITLE
feat: recommended annotation

### LINKED ISSUE #877: Best Practice:  feed_info.txt should be provided (WARNING)
Lack of  `feed_info.txt` should all trigger a WARNING because the file is required in the GTFS Best Practices. 

Should probably mirror or extend [MissingRequiredFileNotice](https://github.com/MobilityData/gtfs-validator/blob/master/RULES.md#MissingRequiredFileNotice) to allow for a parameter to indicate `SeverityLevel.WARNING` 

https://github.com/MobilityData/gtfs-validator/blob/52434454b747bb4085038f27563e98fee6018ca5/core/src/main/java/org/mobilitydata/gtfsvalidator/notice/MissingRequiredFileNotice.java#L27-L29

### PR DESCRIPTION
**Pull request by KClough**
**Summary:**

Related to issue #885

Adds a new annotation for recommended fields

**Expected behavior:** 

Provides additional functionality around the `recommended` annotation. It can now be applied to fields. 

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
~- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)~

@isabelle-dr per [your request](https://github.com/MobilityData/gtfs-validator/pull/1123#pullrequestreview-958240912), this pull request contains just the annotation. I don't think a screenshot makes sense in this scenario.

### CODE REVIEW

File: core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java [hunk 1]
```diff
@@ -107,12 +110,16 @@ public boolean checkRowLength() {
...
   @Nullable
-  public String asString(int columnIndex, boolean required) {
+  public String asString(int columnIndex, FieldLevelEnum level) {
     String s = row.asString(columnIndex);
-    if (required && s == null) {
+    if (level == FieldLevelEnum.REQUIRED && s == null) {
       noticeContainer.addValidationNotice(
           new MissingRequiredFieldNotice(
               fileName, row.getRowNumber(), header.getColumnName(columnIndex)));
+    } else if (level == FieldLevelEnum.RECOMMENDED && s == null) {
```
Comment (maximearmstrong) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:119]: If possible, this could also go in another PR specifically for the notices.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java [hunk 2]
```diff
@@ -341,7 +347,7 @@ public enum NumberBounds {
    * parsing failed.
    *
    * @param columnIndex index of the column to parse
-   * @param required whether the value is required according to GTFS
+   * @param level whether the value is required or recommended according to GTFS
```
Comment (maximearmstrong) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:350]:
```suggestion
   * @param level whether the value is required, recommended or optional according to GTFS
```

File: core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java [hunk 3]
```diff
@@ -376,14 +382,14 @@ private <T> T parseAsType(
    * error is emitted, the value is considered invalid.
    *
    * @param columnIndex index of the column to parse
-   * @param required whether the value is required according to GTFS
+   * @param level whether the value is required or recommended according to GTFS
```
Comment (maximearmstrong) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:385]:
```suggestion
   * @param level whether the value is required, recommended or optional according to GTFS
```

File: core/src/test/java/org/mobilitydata/gtfsvalidator/notice/NoticeContainerTest.java [hunk 4]
```diff
@@ -51,6 +51,18 @@ public void exportNotices() {
                 + "\"message\":\"Index 0 out of bounds\"}]}]}");
   }
 
+  @Test
+  public void exportRecommendedFileNotices() {
```
Comment (maximearmstrong) [inline core/src/test/java/org/mobilitydata/gtfsvalidator/notice/NoticeContainerTest.java:55]: This could also be in the other PR for the notices.

File: core/src/test/java/org/mobilitydata/gtfsvalidator/testgtfs/GtfsStopTableContainer.java [hunk 5]
```diff
@@ -58,4 +58,9 @@ public Optional<GtfsStop> byPrimaryKey(String id, String subId) {
   public boolean isRequired() {
     return true;
   }
+
+  @Override
+  public boolean isRecommended() {
```
Comment (maximearmstrong) [inline core/src/test/java/org/mobilitydata/gtfsvalidator/testgtfs/GtfsStopTableContainer.java:63]: I'm not sure if I'm missing something here, but is there any reason why we would need this for a test related to `GtfsStop`? Also, there seems to be `no usages found` according to IntelliJ.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedInfoSchema.java [hunk 6]
```diff
@@ -38,16 +40,21 @@ public interface GtfsFeedInfoSchema extends GtfsEntity {
...
+  @Recommended
   String feedVersion();
 
+  @Recommended
   @FieldType(FieldTypeEnum.EMAIL)
   String feedContactEmail();
 
+  @Recommended
   @FieldType(FieldTypeEnum.URL)
   String feedContactUrl();
```
Comment (maximearmstrong) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedInfoSchema.java:59]: According to the [spec](https://gtfs.org/schedule/reference/#feed_infotxt), this fields are optional, not recommended. I understand though that the file is recommended. @isabelle-dr What is your input on this? Should fields included in a recommended file be recommended although the spec says they are optional? I think they should stay optional, but I would be happy to hear other thoughts on this.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableLoaderGenerator.java [hunk 7]
```diff
@@ -436,6 +447,11 @@ private MethodSpec generateLoadMissingFileMethod() {
                 classNames.tableContainerTypeName(),
                 classNames.tableContainerTypeName(),
                 TableStatus.class)
+            .beginControlFlow("if (isRecommended())")
+            .addStatement(
+                "noticeContainer.addValidationNotice(new $T(gtfsFilename()))",
+                MissingRecommendedFileNotice.class)
+            .endControlFlow()
```
Comment (maximearmstrong) [inline processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableLoaderGenerator.java:454]: I think this could be added in the other PR, but everything else in this class could stay.

**Review by maximearmstrong — CHANGES_REQUESTED** [review]
Thank you for your contribution @KClough ! This is great work. Overall, if possible, it may be better to split this PR in two so that the notices are added in a second one. Otherwise, only a few comments before approval.

File: core/src/test/java/org/mobilitydata/gtfsvalidator/testgtfs/GtfsStopTableContainer.java (hunk 5 above)
Comment (KClough) [reply core/src/test/java/org/mobilitydata/gtfsvalidator/testgtfs/GtfsStopTableContainer.java:63]: I believe I hit some errors where the existing tests were failing without this logic, I will double check to confirm.

**Comment by KClough** [comment]
Sorry, it looks like I made a mistake when separating the PRs, I meant for this one to just have the annotation logic. I'll take another pass at it.

**Comment by maximearmstrong** [comment]
No worries @KClough! I'll take a look at it right after.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedInfoSchema.java (hunk 6 above)
Comment (maximearmstrong) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedInfoSchema.java:59]: This is related to #885.

**Comment by isabelle-dr** [comment]
Hi @KClough, thanks for splitting the PRs, this makes it simpler :) 

Here is what I understand from the opened PRs:
- PR #1149 (this one) to add a @Recommended annotation in the validator
- PR #1151 which closes issue #877 to add the `missing_required_file` notice, which will apply (for now) only to the`feed_info.txt` file, but that could be used for other files when the Best Practices evolves.
- PR #1152 which closes issue #885 to add the `missing_required_field` notice, which will apply (for now) only to the `feed_info.txt` fields mentioned in the issue. Again, we can then use this notice for other fields recommended by the best practices or spec.

If this is accurate, here are some changes to do, in order to make it easier when we look back at these PRs in the future:
1. Can you please update the description of this PR to remove the link to issue #887, and update the "Expected behaviour section"?
2. Remove the @recommended in the `GTFSFeedInfoSchema` file in this PR, to have it only in PR #1152

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedInfoSchema.java (hunk 6 above)
Comment (maximearmstrong) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedInfoSchema.java]: Perfect, thank you!

File: core/src/test/java/org/mobilitydata/gtfsvalidator/testgtfs/GtfsStopTableContainer.java (hunk 5 above)
Comment (KClough) [reply core/src/test/java/org/mobilitydata/gtfsvalidator/testgtfs/GtfsStopTableContainer.java:63]: Confirmed, I get this error when running `gradlew test` without this change. 

```
> Task :core:compileTestJava FAILED
/Users/kclough/projects/jarvus/gtfs-validator/core/src/test/java/org/mobilitydata/gtfsvalidator/testgtfs/GtfsStopTableContainer.java:26: error: GtfsStopTableContainer is not abstract and does not override abstract method isRecommended() in GtfsTableContainer
public class GtfsStopTableContainer extends GtfsTableContainer<GtfsStop> {
       ^
1 error

FAILURE: Build failed with an exception.
```

File: core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java (hunk 1 above)
Comment (maximearmstrong) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:119]: I believe we will need to remove the `else if` statement here as the `MissingRecommendedFieldNotice` is not yet implemented in this PR. It should be in #1157.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/Recommended.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * Adds a validation that the field or a file is recommended.
+ *
+ * <p>Example.
+ *
+ * <pre>
+ *   @Recommended
+ *   {@literal @}GtfsTable(value = "feed_info.txt", singleRow = true) {
+ *     ...
+ *   }
```
[the remark this answers is not in the document]
Comment (maximearmstrong) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/Recommended.java:33]: We would need this before we merge the PR.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java (hunk 1 above)
Comment (maximearmstrong) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:119]: Actually, most of the checks failing in this PR are because of this - we are importing a class that does not exist in the import section.

File: core/src/test/java/org/mobilitydata/gtfsvalidator/testgtfs/GtfsStopTableContainer.java (hunk 5 above)
Comment (maximearmstrong) [reply core/src/test/java/org/mobilitydata/gtfsvalidator/testgtfs/GtfsStopTableContainer.java:63]: Ah, right, it has to be implemented because the class inherits the abstract method. All good.

**Comment by maximearmstrong** [comment]
Than you for the changes @KClough! 

I believe we are missing a few things before we can merge:
- there are two unresolved items above - I believe one of these will fix most of the checks failing in this PR. 
- the code format workflow fails. Did you format the code using `gradle googleJavaFormat` or `gradle goJF` before pushing your code?
- The branch is out-of-date with the `master` branch. We will need it to be up-to-date to be able to merge.

**Comment by maximearmstrong** [comment]
@bdferris-v2, I've added you as a reviewer for this PR. If you get a chance to look at it, your feedback is welcome :)

File: core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java [hunk 9]
```diff
@@ -157,39 +159,38 @@ public String asEmail(int columnIndex, boolean required) {
    *     unknown, only phone number starting by "+" are validated.
    */
   @Nullable
-  public String asPhoneNumber(int columnIndex, boolean required) {
-    return asValidatedString(columnIndex, required, fieldValidator::validatePhoneNumber);
+  public String asPhoneNumber(int columnIndex, FieldLevelEnum level) {
```
Comment (maximearmstrong) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:162]: We will need to change the doc string here. I think it makes the Task :aggregateJavadoc to fail [here](https://github.com/MobilityData/gtfs-validator/runs/6619597119?check_suite_focus=true) and [here](https://github.com/MobilityData/gtfs-validator/runs/6619597193?check_suite_focus=true).

File: core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/Recommended.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import java.lang.annotation.Target;
+
+/**
+ * Adds a validation that the field or a file is recommended.
+ *
+ * <p>Example.
+ *
+ * <pre>
+ *   {@literal @}GtfsTable(value = "feed_info.txt", singleRow = true) {
+ *   @Recommended
```
Comment (maximearmstrong) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/Recommended.java:31]: My bad, there was an error in my previous suggestion. It makes the Task :aggregateJavadoc to fail [here](https://github.com/MobilityData/gtfs-validator/runs/6619597119?check_suite_focus=true) and [here](https://github.com/MobilityData/gtfs-validator/runs/6619597193?check_suite_focus=true).
```suggestion
 *   {@literal @}Recommended
```

**Review by maximearmstrong — COMMENTED** [review]
2 things that should fix the checks that are failing right now.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java [hunk 11]
```diff
@@ -52,8 +53,9 @@
  */
 public class RowParser {
 
-  public static final boolean REQUIRED = true;
-  public static final boolean OPTIONAL = false;
+  public static final FieldLevelEnum OPTIONAL = FieldLevelEnum.OPTIONAL;
+  public static final FieldLevelEnum RECOMMENDED = FieldLevelEnum.RECOMMENDED;
+  public static final FieldLevelEnum REQUIRED = FieldLevelEnum.REQUIRED;
```
Comment (bdferris-v2) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:58]: Are these used anywhere?

Comment (bdferris-v2) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:58]: Never mind.  Ignore this.

**Review by bdferris-v2 — COMMENTED** [review]
I've got a high-level macro comment that may just highlight my ignorance of how this code works, but let me give it a shot.

I feel like there is a model mismatch in how you are treating Required vs Recommended vs Optional in this PR.  On one hand, you've got the file and field descriptors, which have separate booleans for isRequired() and isRecommended().  On the other hand, you've introduced a new FieldLevelEnum for passing into RowParser specifying the acceptance level for a given field.  This PR, as best I can tell, doesn't actually update the logic in TableLoaderGenerator for setting that FieldLevelEnum when calling RowParser methods, instead relying on the existing isRequired() method and switching between REQUIRED or OPTIONAL, but never actually passing in RECOMMENDED.  Maybe that's a function of the way you split up the PR into parts?

So at a tactical level, I think it would be good to update that logic in this PR.

But more generally, it made me wonder how this should actually be modeled overall.  So I have questions like:

1) Are all Required fields also Recommended by implication?
2) Should the boolean isRequired() and isRecommended() methods in the descriptors be refactored to use FieldLevelEnum directly instead as a single state?

**Comment by maximearmstrong** [comment]
@bdferris-v2 Very good point, I hadn't noticed that this logic implementation was missing in `TableLoaderGenerator`. I also think it should be updated in this PR.

For your questions

1. This is a good question. A field is Required because the specification says it is required. A field is Recommended because the best practices say that the field "should" be there - even if "must" is used in the best practices, we consider it is recommended. Then yes, most probably we can say that Required fields are Recommended by implication. We would not want notices to be created for both though - the highest requirement level reached for a field is the one considered. @isabelle-dr Do you have thoughts on this?

2. Now that you mention it, I think it could be the best. It seems that the `required()` method in `GtfsFieldDescriptor` is only used in `TableLoaderGenerator` to generate the `load()` and `getRequiredColumnNames()` methods of the table loaders. Most likely using Enum values would be better. Though, would we keep the `isRecommended()` and `isRequired()` methods for the file descriptors as is?

**Comment by KClough** [comment]
@bdferris-v2 it looks like I messed up this PR when splitting out notices from annotations. I think we're in good spot now.

@maximearmstrong I believe I've resolved all javaDoc issues. The build step passes locally for me.

Once this PR is merged, I'll rebase my other PRs and resolve any other outstanding issues.

**Comment by bdferris-v2** [comment]
Apologies, but I think my original concern still exists.  Namely, you've updated all the methods in `RowParser` to accept the new `FieldLevelEnum`, but you haven't updated the call site generated in `TableLoaderGenerator` to ever pass in a `RECOMMENDED` value where appropriate.

Specifically, [TableLoaderGenerator L331](https://github.com/MobilityData/gtfs-validator/blob/master/processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableLoaderGenerator.java#L331):

```
"rowParser.$L($L, $T.$L$L)",
    gtfsTypeToParserMethod(field.type()),
    fieldColumnIndex(field.name()),
    RowParser.class,
    field.required() ? "REQUIRED" : "OPTIONAL",  // <= This line right here
    ...
```

That's the piece I'm still looking for in this PR.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableLoaderGenerator.java [hunk 12]
```diff
@@ -328,7 +329,7 @@ private MethodSpec generateLoadMethod() {
               gtfsTypeToParserMethod(field.type()),
               fieldColumnIndex(field.name()),
               RowParser.class,
-              field.required() ? "REQUIRED" : "OPTIONAL",
+              field.required() ? "REQUIRED" : (field.recommended() ? "RECOMMENDED" : "OPTIONAL"),
```
Comment (bdferris-v2) [inline processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableLoaderGenerator.java:332]: Apologies for the nitpick, but I think this would be more straightforward if you referenced `FieldLevelEnum.class` directly here, instead of indirectly referencing the value via RowParser.  Aka replace the value on line 331.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java (hunk 11 above)
Comment (bdferris-v2) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:58]: You could drop these three lines entirely if you made the suggested change below.

Comment (KClough) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java]: I didn't see a suggested change below, but I believe I understand what you're proposing. This does make it a little easier to follow. It doesn't make sense for `RowParser` to have to expose `FieldLevelEnum` values this way. 

Please see my updated commit. 

Thanks!

**Review by bdferris-v2 — APPROVED** [review]

**Comment by isabelle-dr** [comment]
Thanks for following-up on this @bdferris-v2! 
@maximearmstrong, can you have another look?
I have been following from far, do we need to open an issue to re-iterate the logic for adding the @recommended annotation in a later stage?

**Review by maximearmstrong — APPROVED** [review]
Thank you @KClough and @bdferris-v2! The build passes for me as well. We are ready to merge 🎉 

@isabelle-dr I don't think we need to open an issue to re-iterate the logic for adding the `@recommended` annotation. I think this implementation works well.
